src/ProjectTemplates/Directory.Build.targets

```diff
@@ -0,0 +1,148 @@
+      by all template project reference packages in order to extract the package id and version from the project.
+    -->
+    <MSBuild
+      Projects="%(TemplatePackageVersion.Project)"
+      Condition="'%(TemplatePackageVersion.Project)' != '' and '%(TemplatePackageVersion.Version)' == ''"
+      Targets="GetProjectPackageVersion">
+        <!-- Each item has a its PackageId as the ItemSpec, plus Version and Project properties -->
+        <Output TaskParameter="TargetOutputs" ItemName="_ProjectVersions" />
+    </MSBuild>
+
+    <ItemGroup>
+      <!-- Resolved items where a Version was specified -->
+      <_ResolvedTemplatePackageVersion
+        Include="@(TemplatePackageVersion)"
+        Condition="'%(TemplatePackageVersion.Version)' != ''">
+        <Version>%(TemplatePackageVersion.Version)</Version>
+        <Project>%(TemplatePackageVersion.Project)</Project>
+        <VersionResolvedFromProject>false</VersionResolvedFromProject>
+      </_ResolvedTemplatePackageVersion>
+
+      <!-- Resolved items where the Version was resolved from its project version -->
+      <_ResolvedTemplatePackageVersion Include="%(_ProjectVersions.Identity)">
+        <Version>%(_ProjectVersions.Version)</Version>
+        <Project>%(_ProjectVersions.Project)</Project>
+        <VersionResolvedFromProject>true</VersionResolvedFromProject>
+      </_ResolvedTemplatePackageVersion>
+    </ItemGroup>
+  </Target>
+
+  <!-- Invoke this target before 'Build' on any project that imports this target file -->
+  <Target Name="_GenerateTemplateContent"
+    BeforeTargets="Build"
+    DependsOnTargets="_ResolveTemplatePackageVersions">
+    <PropertyGroup>
+      <!-- Generated template content files are grouped by template package id to avoid collisions -->
+      <GeneratedContentDir>$([System.IO.Path]::Combine($(ArtifactsTemplateGeneratedContentDir), $(PackageId)))</GeneratedContentDir>
+    </PropertyGroup>
+
+    <!--
+      Template packages define <TemplateContent> items to pre-process for token replacements,
+      including <TemplatePackageVersion> items that define package version replacements.
+
+      All template content files are output into the project's <Content> to ensure the output
+      files are added to the template package.
+
+      The <TemplatePackageVersion> items have been resolved to <_ResolvedTemplatePackageVersion> items
+      to populate any Version properties needed from the Project properties.
+    -->
+    <GenerateTemplateContentFiles
+      SourceFiles="@(TemplateContent)"
+      GeneratedContentDir="$(GeneratedContentDir)"
+      PackageReferences="@(_ResolvedTemplatePackageVersion)">
+      <Output TaskParameter="GeneratedContent" ItemName="Content" />
+    </GenerateTemplateContentFiles>
+  </Target>
+
+  <UsingTask TaskName="GenerateTemplateContentFiles" TaskFactory="RoslynCodeTaskFactory" AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.Core.dll">
```

Comment:
Can you use https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.Build.Tasks.Templating/src/GenerateFileFromTemplate.cs instead?  We try to avoid the use of custom tasks like this if something already exists, especially those that use RoslynCodeTaskFactory as it's difficult to debug and perf is bad.